apiserver/facades/agent/uniter/networkinfo.go

```diff
@@ -0,0 +1,458 @@
+	relEgress := state.NewRelationEgressNetworks(n.st)
+	egressSubnets, err := relEgress.Networks(rel.Tag().Id())
+	if err != nil && !errors.IsNotFound(err) {
+		return "", nil, nil, errors.Trace(err)
+	} else if err == nil {
+		egress = egressSubnets.CIDRS()
+	} else {
+		egress = n.defaultEgress
+	}
+
+	boundSpace, err = n.spaceForBinding(binding)
+	if err != nil && !errors.IsNotValid(err) {
+		return "", nil, nil, errors.Trace(err)
+	}
+
+	fallbackIngressToPrivateAddr := func() error {
+		// TODO(ycliuhw): lp-1830252 retry here once this is fixed.
+		address, err := n.unit.PrivateAddress()
+		if err != nil {
+			logger.Warningf("no private address for unit %q in relation %q", n.unit.Name(), rel)
+		} else if address.Value != "" {
+			ingress = append(ingress, address)
+		}
+		return nil
+	}
+
+	// If the endpoint for this relation is not bound to a space, or
+	// is bound to the default space, we need to look up the ingress
+	// address info which is aware of cross model relations.
+	if boundSpace == corenetwork.DefaultSpaceName || err != nil {
+		_, crossModel, err := rel.RemoteApplication()
+		if err != nil {
+			return "", nil, nil, errors.Trace(err)
+		}
+		if crossModel && (n.unit.ShouldBeAssigned() || pollPublic) {
+			address, err := pollForAddress(n.unit.PublicAddress)
+			if err != nil {
+				logger.Warningf(
+					"no public address for unit %q in cross model relation %q, will use private address",
+					n.unit.Name(), rel,
+				)
+			} else if address.Value != "" {
+				ingress = append(ingress, address)
+			}
+			if len(ingress) == 0 {
+				if err := fallbackIngressToPrivateAddr(); err != nil {
+					return "", nil, nil, errors.Trace(err)
+				}
+			}
+		}
+	}
+
+	if len(ingress) == 0 {
+		if n.unit.ShouldBeAssigned() {
+			// We don't yet have an ingress address, so pick one from the space to
+			// which the endpoint is bound.
+			machineID, err := n.unit.AssignedMachineId()
+			if err != nil {
+				return "", nil, nil, errors.Trace(err)
+			}
+			machine, err := n.st.Machine(machineID)
+			if err != nil {
+				return "", nil, nil, errors.Trace(err)
+			}
+			sp, err := n.lookupSpaces(boundSpace)
+			if err != nil {
+				return "", nil, nil, errors.Trace(err)
+			}
+			networkInfos := machine.GetNetworkInfoForSpaces(sp)
+			// The binding address information based on link layer devices.
+			for _, nwInfo := range networkInfos[boundSpace].NetworkInfos {
+				// We need to construct sortable addresses from link-layer
+				// devices, which unlike machine addresses do not have this
+				// information. We can at least ensure that fan addresses will
+				// be sorted after other addresses.
+				scope := corenetwork.ScopeCloudLocal
+				if strings.HasPrefix(nwInfo.InterfaceName, "fan-") {
+					scope = corenetwork.ScopeFanLocal
+				}
+
+				for _, addr := range nwInfo.Addresses {
+					ingress = append(ingress, corenetwork.NewScopedSpaceAddress(addr.Address, scope))
+				}
+			}
```

Comment:
This looks identical to a chunk above except it doesn't call Sort, which 
a) We should probably Sort
b) We should refactor into a shared function

unless I'm missing why they aren't the same.

Reply:
This is a pain because on is a `[]network.NetworkInfo` and the other is `[]params.NetworkInfo` I'm hoping this is temporary in any case based on potentially getting scope information via the new to-be-written InstancePoller into the `ip.addresses` collection.

Reply:
I extracted it; ended up being simple.